### ISSUE TITLE
Warn about too few candidate passwords

### ISSUE THREAD

**Issue by solardiz** [magnumripper/JohnTheRipper#3386]
It is very common for users to run -opencl and -ztex formats on small wordlists, or even to try "single crack" mode, not realizing they're leaving most hardware resources idle. Maybe we should have a check in shared code (in cracker.c?) where we'd print a warning when the count passed to `crypt_all()` is below `min_keys_per_crypt` (include both numbers in the warning message, e.g. "Warning: only %u candidate passwords left, which is below the minimum of %u needed to utilize the available hardware resources"). Maybe print it the first time only, or maybe print it each time the count is different from what was seen previously (such that there would be many of those in "single crack" mode, consistent with how extremely bad the resulting performance is, but only one for the last batch in wordlist mode as the count will stay the same across all salts), maybe up to a certain number of these messages printed (then print that "further warnings of this type are suppressed" or similar). Maybe the verbosity setting should affect whether/when this warning is printed.

Alternatively, we could implement this inside the -ztex formats for now, printing a warning when some of the hardware cores are left idle. But I think the problem is more generic.

**Comment by magnumripper** [comment]
While on that page we could also want to make "batch mode" different when format is ZTEX or OpenCL. It could be confusing though.

Just thinking loud, another alternative would be to have batch mode enforce CPU format for single mode, so that `./john hashes -form:nt-opencl` would start with single mode on CPU despite the format `-opencl` option, then continue with wordlist and incremental using OpenCL as requested. This too could be confusing but pretty handy (and it could be explained with some informative message).

FWIW some early versions of OpenCL support for RAR3 was built into the existing CPU format (iirc there was only `-format:rar` and it would pick OpenCL if applicable). For single mode and small batches it would simply call the CPU version of `crypt_all()` instead (and I think the decision was made per call so a wordlist run would often use OpenCL for all of the session except the very last batch). It worked just fine but I abandoned the idea anyway because it was too confusing - and made it hard to run a CPU session alongside an OpenCL one in a controlled manner. But I digress. Nowadays we have many OpenCL formats that are so different from their CPU counterparts it wouldn't be possible at all to just call "the other" `crypt_all()` transparently.

**Comment by magnumripper** [comment]
PoC code. This seems to induce a 2% speed drop for NT. Not sure what would be a good default for the `max_warn` variable? It could be a params.h or even john.conf setting if we wanted to. This code currently lacks `john_main_process` as a condition for printing (and maybe that's correct).
```diff
diff --git a/src/cracker.c b/src/cracker.c
index b804fa956..38fe2a3d9 100644
--- a/src/cracker.c
+++ b/src/cracker.c
@@ -820,6 +820,20 @@ static int crk_password_loop(struct db_salt *salt)
 	if (fp_fix_state)
 		fp_fix_state();

+	if (options.verbosity >= VERB_DEFAULT &&
+	    crk_key_index < crk_params->min_keys_per_crypt) {
+		static int last, num_warns;
+		const int max_warn = 10;
+
+		if (crk_key_index != last && num_warns < max_warn) {
+			last = crk_key_index;
+			fprintf(stderr, "Warning: only %d candidates left, minimum %d needed for performance\n",
+			        crk_key_index, crk_params->min_keys_per_crypt);
+			if (++num_warns == max_warn)
+				fprintf(stderr, "Further warnings of this type will be suppressed\n");
+		}
+	}
+
 	count = crk_key_index;
 	match = crk_methods.crypt_all(&count, salt);
 	crk_last_key = count;
```

**Comment by solardiz** [comment]
Thanks, magnum!

> This seems to induce a 2% speed drop for NT.

That's weird. This suggests that maybe our `max_keys_per_crypt` for NT needs to be increased.

> Not sure what would be a good default for the `max_warn` variable? It could be a params.h or even john.conf setting if we wanted to.

Right, but hard-coded 10 is fine for the initial commit.

> This code currently lacks `john_main_process` as a condition for printing (and maybe that's correct).

Yes, that's correct, but we need to include node number in the messages (perhaps as "%d: Warning: ..." when there's more than one node). Otherwise, the messages will be confusing (especially the one about "Further warnings ... suppressed" when it's nevertheless followed by a further warning from another process).

```c
if (options.verbosity >= VERB_DEFAULT &&
    crk_key_index < crk_params->min_keys_per_crypt) {
```

These checks should be swapped (check `crk_key_index` first) for better performance in the typical case (when there are enough candidates yet the verbosity level is sufficient for this warning to be potentially printed).

**Comment by solardiz** [comment]
I wonder if we might sometimes receive spurious warnings with on-device mask mode. Should we possibly be checking `*count` as possibly-modified by `crypt_all()` instead of `crk_key_index` before `crypt_all()`? Originally, `min_keys_per_crypt` was meant to apply to the count input to `crypt_all()`, but common sense suggests that for this check the possibly-modified `*count` is more relevant.

**Comment by magnumripper** [comment]
We should be able to use `crk_key_index * mask_int_cand.num_int_cand` for a correct figure even before calling `crypt_all()`. I'll experiment with it when I get some time.

**Comment by solardiz** [comment]
Maybe:

```c
if (crk_key_index < crk_params->min_keys_per_crypt &&
    crk_key_index * mask_int_cand.num_int_cand < crk_params->min_keys_per_crypt &&
    options.verbosity >= VERB_DEFAULT)
```

so that we avoid the multiplication in the typical case.

We may also change `crk_params` from a pointer to a copy of the structure (like we do for `crk_methods`) to avoid the pointer indirection, but we'd need to be careful to do that after any `init()` and `reset()` calls we might make (not before). And this would need to be a separate commit.

**Comment by magnumripper** [comment]
On another note, most OpenCL formats set min. keys per crypt to LWS, which is often as low as 32 or 64. So we sometimes wont get any warning even though running suboptimal.

**Comment by solardiz** [comment]
"most OpenCL formats set min. keys per crypt to LWS" - maybe they shouldn't. Maybe they should set it either to whatever is needed to make at least some use of all hardware, or even to `max_keys_per_crypt`.

Regardless, ideally we'd have two types of warnings here: for below max and for below min (should be a nested if to skip that check most of the time). Should use different wording there. Also, in the warning for below min should say "ideally %d" and give max there.

**Comment by magnumripper** [comment]
On second thought I believe GPU-side mask multiplier (or the modifed *count) is actually irrelevant because the min/max kpc's already assume the multiplier.

Also, I'm leaning towards *only* looking at max kpc. I tried checking both, using different messages, and I'm sure the typical user would only get confused.

Stepping back a little though, when will this situation actually happen?
1. Running single mode using GPU/FPGA. This could be warned for by other means.
2. At length-switch in iterated mask mode using GPU-side mask. This is a special case where we'd probably be better off not warning. Also, it's typically salt-less (at least on GPU, not sure about FPGA?).
3. At the last crypt of a run, in any mode. For single-salt or few salts this one can be ignored and maybe we shouldn't even print it (we'd just confuse people). However, with many salts this case can get similar to the next case:
4. Running eg. a small wordlist with no (or few) rules and LOTS of salts. After single, this is the worst case.

Perhaps we should take salt count into the equation somehow.

**Comment by solardiz** [comment]
Posting whatever comments I readily have:

"the min/max kpc's already assume the multiplier" - sure, but `crk_key_index` does not! That's precisely why I suggested looking at the maybe-modified `*count` (and am also OK with looking at `crk_key_index * mask_int_cand.num_int_cand`).

"Also, I'm leaning towards only looking at max kpc." - this wouldn't be helpful enough e.g. to Aleksey's use of ZTEX boards in the recent CMIYC. In there, a common task was to use the smallest wordlist size that uses the hardware fully, ignoring the last 1% or whatever of performance maybe achievable by going further from min to max. Arguably, the bcrypt-ztex format should have lowered max to be closer to or even same as min at bcrypt cost this high, but I'm sure there are other similar examples as well.

**Comment by magnumripper** [comment]
> "the min/max kpc's already assume the multiplier" - sure, but crk_key_index does not! That's precisely why I suggested looking at the maybe-modified *count (and am also OK with looking at crk_key_index * mask_int_cand.num_int_cand).

What I meant is that min/max `kpc` as well as `crk_key_index` and LWS/GWS all are **without** the multiplier. Only the returned `pcount` is larger. For example, running `?d?d?d?d?d` may mean we have a  `mkpc` (and `crk_key_index`, and GWS) of 100 and a `num_int_cand` of 1000.

**Comment by solardiz** [comment]
Yes, all of these are without the multiplier, but aren't we setting mkpc to whatever high number of candidates (post-mask, if a mask is used) is needed to optimally use the hardware? In your example, if we ended up having `crk_key_index * num_int_cand` at 100k, wouldn't mkpc likely be 100k as well (or otherwise we wouldn't have needed this product to be so high, and likely wouldn't have made it so high)?

Regarding your list of cases where the warning would possibly appear:

1. Yes, we can have separate warnings for `--single` mode, but until we do this generic warning is possibly better than nothing.
2. The length-switching example is interesting, and relevant on FPGA for slow, salted hashes. Perhaps we should print a different warning when this issue occurs, and perhaps we shouldn't suppress it even after it's been printed many times (the number of length switches can't be greater than the number of lengths, so this provides a natural limit on the number of times the warning will be printed). A similar example is switching between passes in batch mode.
3. Yes, it can happen for the last chunk, and I'd like the warning printed. But perhaps only when processing of that chunk for all salts is expected to take more than, say, 1 second. This is a function of hash type, its cost parameters, and salt count. Tricky.
4. Yes, it should happen when running a small wordlist against lots of salts or/and some especially slow (non-)hash. This was the primary case I was thinking of when I created this issue.

As you point out, 3 and 4 are related - e.g., if a wordlist is of 5k lines whereas mkpc is 4k, things can get very non-optimal for one out of two chunks (the second one will have only 1k lines out of the optimal 4k, possibly leaving 3/4 of hardware idle), and this will have very significant effect on overall performance of that run.

**Comment by solardiz** [comment]
OK, I guess I misunderstood how things worked with on-device mask as it relates to mkpc adjustment.

**Comment by solardiz** [comment]
FWIW, running `-mask='?a?a?a?a?a?a?a'` against one sha512crypt hash (default round count) with sha512crypt-ztex gives me this in john.log:

```
- Candidate passwords will be buffered and tried in chunks of 16384  
...
- Requested internal mask factor: 16640, actual now 9025
```

**Comment by magnumripper** [comment]
If I saw those lines from an OpenCL format I would assume we'd do `crypt_all(16384)` and get back a `count` of 147,865,600. I can't really see how a ztex format could work differently since we have no alternative core code path.

**Comment by claudioandre-br** [comment]
A real example using `raw-SHA512-opencl` and `-mask='?a?a?a?a?a?a?a' ` on super.

GWS=3072 x mask factor=9025 => 27M candidates to handle on `every crypt_all()` on GPU (good value).

There is no way one can complain about too few candidates. 9025 candidates will be produced by mask for **each** "plaintext".
****

Mask breaks the work to be done in `-mask='?a?a?a?a?a?a?a' `:
- `?a` => 95 candidates (very little)
- `?a?a` => 95x95(9025) candidates (good)
- `?a?a?a` => 95x95x95 candidates (too much)

BTW: 9025 is **NOT** the best value for a fast GPU format since it can't hide all latencies. A better mask would be ?l?l?l (17K candidates).

Any chance you're not happy with the latencies?

**Comment by solardiz** [comment]
"Any chance you're not happy with the latencies?" - no, JtR is actually quite responsive when running with that mask against sha512crypt-ztex, which is why I think it does _not_ actually produce the ~148M candidates per `crypt_all()` call as magnum calculated (that would be taking like an hour on one ZTEX board). So I think the mask multiplier's behavior on ZTEX is different than in OpenCL, and I'd like Denis to confirm and explain that.

**Comment by claudioandre-br** [comment]
> it's weird to have on-device mask for a slow hash like that

No, it is not.
****
1. Well, In fact, I'm not talking about slow or fast hashes. 
2. I would say you are unhappy with
```
Requested internal mask factor: 16640, actual now 9025
```

So, what I said is: since mask factor is 9025, a wordlist of 16 entries will result in more than 140K candidates generation on device (for ONE crypt_all() call). 

So, I agree with magnum in https://github.com/magnumripper/JohnTheRipper/issues/3386#issuecomment-423772550. The mask factor can't hurt, on the contrary.

**Comment by solardiz** [comment]
> So, what I said is: since mask factor is 9025, a wordlist of 16 entries will result in more than 140K candidates generation on device (for ONE crypt_all() call).

Like I said, in sha512crypt-ztex this is apparently _not_ the case.

**Comment by magnumripper** [comment]
> Maybe max_keys_per_crypt is lowered at some later point 

I was just going to write the same. I can't see any other explanation.

**Comment by Apingis** [comment]
Regarding mask handling for slow hashes in ZTEX formats. 
- There's no autotune, the approximate performance is known and is hardcoded.
- 1st it calculates ```device_bitstream.candidates_per_crypt``` (number of candidates for each FPGA). That's based on ```TargetRounds``` and is a multiple of ```device_bitstream.min_keys``` ("Min. number of keys for effective device utilization") - the number of threads in each FPGA.
- It turns mask on if there's 50+ hashes/thread/FPGA. Without this there would be performance loss because mask factor is typically not a multiple of number of threads. 
- On ```reset()``` actual mask factor is known, it sets keys_per_crypt equal to number of candidates for each FPGA, divided by mask factor, multiplied by total number of FPGAs.

Example: it runs ```--form=sha512crypt-ztex --mask=?a?a?a``` with default rounds:
- Hardcoded value is: if hashes are of 256K+ rounds then there must be 1 candidate/thread/crypt_all. For rounds=5000 it results in 52 candidates/thread/crypt_all. The value is multiplied by the number of threads in FPGA: 52 * 160 = 16640. This is the requested mask factor.
- On ```reset()``` it reports mask factor to be 9025. It computes keys_per_crypt: ```keys_per_crypt = jtr_bitstream->candidates_per_crypt / mask_num_cand()```. So keys_per_crypt==1. If mask factor was 18050 then keys_per_crypt could be 2.
- Each thread is to compute ceil(9025/160) = 57 candidates. If there were no onboard mask generation then it would have computed 57*160 = 9120 candidates, so performance loss is about 1%.
- Then it multiplies keys_per_crypt by the number of FPGA.

> So, what I said is: since mask factor is 9025, a wordlist of 16 entries will result in more than 140K candidates generation on device (for ONE crypt_all() call).

With 1 board (4 FPGA) it would set min/max keys_per_crypt=4 (~36K candidates generated). That's less than 1s response time.

**Comment by magnumripper** [comment]
So after `reset()` all these figures should be set correctly, that's good and it's in harmony with the OpenCL formats.

That log information Solar quoted is printed **after** the reset so I still don't quite understand how it could end up like that.

**Comment by solardiz** [comment]
Thanks for the response, @Apingis! This sounds like development documentation we could want to have somewhere in the tree (please feel free to suggest a directory/file to put it in) or on john-dev, not only on this GitHub issue (since GitHub is a third-party, I assume GitHub content is not under our control and might be lost earlier than our project becomes irrelevant).

As you point out, on-device mask may result in performance loss when the mask factor isn't a multiple of the number of hardware threads. Maybe we should warn of this as well (only when the mask is an especially poor fit for the hardware?) and have a way to disable use of on-device mask (for slow hashes only?)

**Comment by magnumripper** [comment]
So I found out why there were more warnings than I expected while testing #3480 in single mode. It's "on purpose". The majority come from this:
```c
	if (keys->count && rule_number - keys->rule > (key_count << 1))
		if (single_process_buffer(salt))
			return 1;
```
@solardiz I assume this is for not delaying early cracks too much, and/or for not letting the "oldest rule still in use" getting to far away. So we have some alternatives, including:
1. We could disable the above code for OpenCL and/or ZTEX formats. Is that reasonable?
2. We could obviously leave things as-is and get the many warnings.
3. We could have a john.conf option choose between the above. If so, what should be the default?

Here's another place that could trigger it although I've yet to see it with some extra debug prints:
```c
	if (!have_words) {
		keys->have_words = 0;
 no_own_words:
		if (keys->count && single_process_buffer(salt))
			return 1;
	}
```
I guess we should leave this as-is and get the warnings.

BTW, both above cases make the message "Warning: Only *n* candidates left(...)" a bit misleading. We may want to rephrase it to something more generic.

Except for those I believe it's only the last crypt per salt that may trigger the warning. Personally I'd like to mute them (or people will flood us with questions), and especially for non-OpenCL/ZTEX but I guess we could make it configurable.

**Comment by magnumripper** [comment]
BTW an alternative for the first (and common) case, for OpenCL only, would be to just modify it a tad so we're not ending up with a count that happens to be eg. a prime number or otherwise unfortunate multiple of LWS, because that would push LWS down to horrible performance. That is, we'd leave it mostly as-is except we'd also wait for keys->count to be a multiple of LWS. Perhaps even wait for it to be a multiple of `(LWS * some constant)`. 

If we go for that I think we can mute the KPC warnings for this very case (as they should be more or less benign then).

**Comment by solardiz** [comment]
@magnumripper You're correct about the rationale for `rule_number - keys->rule > (key_count << 1)`, and it's mostly the second reason you listed: not letting the "oldest rule still in use" getting too far away, as that's how we're able to interrupt/restore single mode sessions at all. If we let "oldest still in use" stay at 0 until very end on any one salt, the whole session would have to be restarted from scratch even if we interrupt it at 99%. This logic breaks down when `key_count` is very large. I guess in your tests it'd only allow to restoring to 0% vs. 50% or something like that. Maybe that's a reason to exclude it when rule count is less than e.g. `4 * min_keys_per_crypt`. But this needs more thought and testing.

I agree the message should be rephrased when in single mode. Maybe "Only n candidates currently buffered for the current salt". That's not more generic, but is even more specific - yet correct.

"People will flood us with questions" means those people would otherwise unknowingly waste their hardware resources. That's not good. Instead of muting, we could have single mode skipped from batch mode (default run with no options) when using OpenCL and ZTEX formats. We could print one warning about that.

Regarding primes and multiples of LWS, I'm not familiar with that but what you wrote feels weird to me. How can the number of candidate passwords possibly "push LWS down to horrible performance"? Aren't we tuning for optimal performance and determine min/max KPC _before_ we even have any candidates?

**Comment by magnumripper** [comment]
> Regarding primes and multiples of LWS, I'm not familiar with that but what you wrote feels weird to me. How can the number of candidate passwords possibly "push LWS down to horrible performance"? Aren't we tuning for optimal performance and determine min/max KPC before we even have any candidates?

What we determine are good local and global worksizes (LWS and GWS). GWS must be a multiple of LWS. Then we set min/max KPC to GWS (possibly after multiplying with a vector width but that's not important for this discussion).

The problem is when we later call `crypt_all()` with a `count` lower than GWS. As long as `count` is still a multiple of LWS, all is fine. But if it's not, we have two options:
1. Set GWS to next LWS multiple *larger* than `count`. That's what we did until a while ago but it's suboptimal and bug prone because we need to transfer more data than actually needed and even so we might run with uninitialized keys for the "extra" ones so we need to add guards or omit optimizations in the kernel code.
2. Set GWS to `count` and LWS to the largest factor of it. We don't need to calculate it but instead use an LWS of NULL and the runtime takes care of it. However, in some cases that ends up with a GWS lower than 32, which is starting to really affect performance, or even to 1 (for a prime `count`) which is really terrible.

Anyway this doesn't seem to be much of a problem with this OpenCL-specific change (similar to the idea I mentioned):
```c
        if (keys->count && rule_number - keys->rule > (key_count << 1))
#if HAVE_OPENCL
        if (ocl_fmt && keys->count % (local_work_size  * ocl_v_width) == 0)
#endif
                if (single_process_buffer(salt))
                        return 1;
```
This will still allow smaller batches for OpenCL but not until they are a multiple of LWS. I'm pretty sure this will be good enough for OpenCL that we can even mute the warnings for this case (still testing though).

For ZTEX, I have no idea though.

**Comment by solardiz** [comment]
The old "Set GWS to next LWS multiple larger than count" approach sounds like the right one to me, although I understand it requires us to have logic to pass a little bit of properly formatted dummy data back and forth. I think we should revert to this approach. Not so much for single mode, but also for other cases where e.g. a wordlist's number of lines wouldn't have a suitable factor for a good LWS, or at least this would commonly be the case for the last portion of a wordlist after having a whole number of `max_keys_per_crypt` sized portions already processed. The "Set GWS to count and LWS to the largest factor of it" approach sounds broken by design to me, if the runtime's handling of it is as bad as you describe.

Meanwhile, I'm fine with the `keys->count % (local_work_size * ocl_v_width) == 0` check. This may mean that "oldest rule still in use" can be stuck at 0 all the time, not allowing for sessions to be restored, but I guess we have to accept that when using OpenCL.

For ZTEX formats, I'd fully skip this piece.

**Comment by magnumripper** [comment]
> mute reporting for all cracking modes

Fixed in 62361fb4a. We only mute if running single mode, and only unless verbosity was bumped.

**Comment by solardiz** [comment]
Note that in single mode with the check revised as you proposed, being a multiple of LWS probably means being exactly LWS, and that means using only 1 out of tens of CUs/SMs in a GPU. Definitely something to warn about, and maybe we need to fully skip that `if (keys->count && rule_number - keys->rule > (key_count << 1))` piece on both OpenCL and ZTEX.

**Comment by magnumripper** [comment]
I didn't see that in my testing. I saw fairly high numbers. I don't have the whole picture but I suspect that between two occasions of this test, a salt might get more than one key added? Then we're likely to miss the LWS "chance" most of the times.

Anyway we could default to skip the whole test for OpenCL as well, but I'd make it configurable.

**Comment by magnumripper** [comment]
This is now merged to main tree. I ended up with no muting of warnings in cracker.c, and with disabling that single-mode low-kpc-crypt for OpenCL and ZTEX.

I'm closing this now, let's open new issues for future problems/features/tweaks.

**Comment by claudioandre-br** [comment]
> The old "Set GWS to next LWS multiple larger than count" approach sounds like the right one to me

To me it sounds like the best/safer approach.

> although I understand it requires us to have logic to pass a little bit of properly formatted dummy data back and forth.

In fact, you only need to clean/initialize OpenCL buffers once (you must provide valid data all the time or "zeroes". A "key" from a previous `crypt_all` call is valid data (key[index] > count && index < GWS); if this "key" is a crack, the `cmp_*` will check upto `count`)

> The "Set GWS to count and LWS to the largest factor of it" approach sounds broken by design to me, if the runtime's handling of it is as bad as you describe.

On the other hand:
* we have not seen people complaining so far. And magnum have tested it.

**Comment by solardiz** [comment]
Regarding "we have not seen people complaining so far", that's no surprise and doesn't mean things are OK. People don't know what performance to expect for tiny wordlists, and whether things can be better than they are or not. Even in team john-users participation in contests, I've seen otherwise-experienced team members unknowingly waste hardware resources by running too small (but highly focused) wordlists against many salted slow hashes on GPU and ZTEX. This is what prompted the creation of this very issue in the first place.

**Comment by magnumripper** [comment]
Do we really need to warn at all when salt count is 1, or even with a hand-full of salts? Sure, performance will be suboptimal but it won't have much impact unless we repeat that for many salts.

Also, do we really need to warn if we get eg. 95% of KPC?

Combining these, how about warning only when count is lower than `min_kpc - min_kpc / salt_count`? Then we'd not warn for single salt but for 10 salts we'd allow 10% underfill and for 100 salts we'd allow only one percent underfill. For even more salts, we'll require a full or very nearly full batch. @solardiz I'm really not expecting you to like that 😆 🤣 but I think it's reasonable. Of course I could make it configurable.

**Comment by magnumripper** [comment]
> Combining these, how about warning only when count is lower than min_kpc - min_kpc / salt_count? Then we'd not warn for single salt but for 10 salts we'd allow 10% underfill and for 100 salts we'd allow only one percent underfill. For even more salts, we'll require a full or very nearly full batch.

After some testing I committed this, and even made it the default (but configurable it is). You need to convince me why we'd warn about single salt.
